### PR TITLE
[Improve][Connector-V2][Doris]Refactor some Doris Sink code as well as support 2pc and cdc

### PR DESCRIPTION
**Pull request by zy-kkk**
## Purpose of this pull request

There were some problems with the existing Doris connector, and I refactor some Doris sink code as well as  support 2pc and cdc

Due to insufficient workflow test resources, all Doris tests cannot be completed, so I disable the test class first. I have successfully tested it locally. Later, I will streamline the docker image so that the workflow test can be successful
image

## Check list

* [x] Code changed are covered with tests, or it does not need tests for reason:
* [x] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [x] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [x] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [x] Update the [`release-note`](https://github.com/apache/incubator-seatunnel/blob/dev/release-note.md).

### CODE REVIEW

File: seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/sink/committer/DorisCommitter.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+            }
+        }
+
+        if (statusCode != HTTP_TEMPORARY_REDIRECT) {
+            throw new DorisConnectorException(
+                    DorisConnectorErrorCode.STREAM_LOAD_FAILED, reasonPhrase);
+        }
+
+        ObjectMapper mapper = new ObjectMapper();
+        if (response.getEntity() != null) {
```
Comment (CalvinKirs) [inline seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/sink/committer/DorisCommitter.java]: Beware of NPE

File: seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/rest/RestService.java [hunk 2]
```diff
@@ -0,0 +1,674 @@
...
+        BufferedReader in =
+                new BufferedReader(
+                        new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8));
+        String line;
+        while ((line = in.readLine()) != null) {
+            result.append(line);
+        }
+        if (in != null) {
+            in.close();
+        }
```
Comment (CalvinKirs) [inline seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/rest/RestService.java]: it's better to used `try catch reource`, and L226 is not null forever

File: seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/rest/RestService.java [hunk 3]
```diff
@@ -0,0 +1,674 @@
...
+    }
+
+    @Deprecated
+    @VisibleForTesting
+    static List<BackendRow> getBackends(DorisConfig dorisConfig, Logger logger)
+            throws DorisConnectorException, IOException {
+        String feNodes = dorisConfig.getFrontends();
+        String feNode = randomEndpoint(feNodes, logger);
+        String beUrl = "http://" + feNode + BACKENDS;
```
Comment (CalvinKirs) [inline seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/rest/RestService.java]: What about abstracting away the url build?

File: seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/sink/committer/DorisCommitter.java (hunk 1 above)
Comment (zy-kkk) [reply seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/sink/committer/DorisCommitter.java]: done

File: seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/rest/RestService.java (hunk 2 above)
Comment (zy-kkk) [reply seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/rest/RestService.java]: done

File: seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/rest/RestService.java (hunk 3 above)
Comment (zy-kkk) [reply seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/rest/RestService.java]: done

**Review by CalvinKirs — DISMISSED** [review]

File: docs/en/connector-v2/sink/Doris.md [hunk 4]
```diff
@@ -9,29 +9,25 @@ The internal implementation of Doris sink connector is cached and imported by st
...
+| fenodes            | string | yes      | -             |
+| username           | string | yes      | -             |
+| password           | string | yes      | -             |
+| table.identifier   | string | yes      | -             |
+| sink.label-prefix  | string | yes      | -             |
+| sink.enable-2pc    | bool   | no       | true          |
+| sink.enable-delete | bool   | no       | false         |
+| doris.config       | map    | yes      | -             |
+
+### node_urls [string]
```
Comment (TyrantLucifer) [inline docs/en/connector-v2/sink/Doris.md]: Means that `fenodes`?

File: seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/sink/writer/DorisSinkState.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (TyrantLucifer) [inline seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/sink/writer/DorisSinkState.java]: `/**` is better

File: seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/sink/writer/DorisSinkWriter.java [hunk 6]
```diff
@@ -0,0 +1,273 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (TyrantLucifer) [inline seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/sink/writer/DorisSinkWriter.java]: The same as above.

File: seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/sink/writer/LabelGenerator.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (TyrantLucifer) [inline seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/sink/writer/LabelGenerator.java]: The same as above

**Comment by TyrantLucifer** [comment]
image

It seems that CI has some problems. Please fix it.

**Comment by CalvinKirs** [comment]
BTW, It is best to add an adapted Doris version

File: docs/en/connector-v2/sink/Doris.md (hunk 4 above)
Comment (zy-kkk) [reply docs/en/connector-v2/sink/Doris.md]: done

File: seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/sink/writer/DorisSinkState.java (hunk 5 above)
Comment (zy-kkk) [reply seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/sink/writer/DorisSinkState.java]: done

File: seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/sink/writer/DorisSinkWriter.java (hunk 6 above)
Comment (zy-kkk) [reply seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/sink/writer/DorisSinkWriter.java]: done

File: seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/sink/writer/LabelGenerator.java (hunk 7 above)
Comment (zy-kkk) [reply seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/sink/writer/LabelGenerator.java]: done

**Review by CalvinKirs — DISMISSED** [review]

File: docs/en/connector-v2/sink/Doris.md [hunk 8]
```diff
@@ -138,5 +124,5 @@ sink {
 
 ### Next version
 
-- [Improve] Change Doris Config Prefix [3856](https://github.com/apache/incubator-seatunnel/pull/3856)
+- [Improve] Refactor some Doris Sink code as well as support 2pc and cdc [4235](https://github.com/apache/incubator-seatunnel/pull/4235)
```
Comment (CalvinKirs) [inline docs/en/connector-v2/sink/Doris.md]: BTW, this is an incompatible change, so we better reflect it in the documentation as well.

**Comment by TyrantLucifer** [comment]
It seems that has some confilcts.

**Review by CalvinKirs — APPROVED** [review]

**Review by TyrantLucifer — APPROVED** [review]
LGTM

File: seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/sink/writer/DorisSinkStateSerializer.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+    @Override
+    public DorisSinkState deserialize(byte[] serialized) throws IOException {
+        try (final ByteArrayInputStream bais = new ByteArrayInputStream(serialized);
+                final DataInputStream in = new DataInputStream(bais)) {
+            final String labelPrefix = in.readUTF();
+            final long checkpointId = in.readLong();
+            return new DorisSinkState(labelPrefix, checkpointId);
+        }
+    }
```
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/sink/writer/DorisSinkStateSerializer.java:48]:
``` java
out.writeUTF(dorisSinkState.getLabelPrefix());
```

 but 

``` java
final String labelPrefix = in.readUTF();
final long checkpointId = in.readLong();
```
Not match.

**Comment by Hisoka-X** [comment]
There some problem, you add support for 2pc, but you don't open 2pc in your e2e config. It cause 2pc can't pass with ci.

**Pull request merged**